### PR TITLE
refactor(execute): expose direct access to process message for transformations

### PR DESCRIPTION
**Pull request by jsternberg**
The `ProcessMessage` method can now be directly implemented by
transformations and the adapter code is present for existing
transformations.

### Done checklist
- [x] docs/SPEC.md updated
- [x] Test cases written

### CODE REVIEW

File: execute/transport.go [hunk 1]
```diff
@@ -263,10 +267,24 @@ type Message interface {
 type MessageType int
 
 const (
+	// RetractTableType is sent when the previous table for
```
Comment (wolffcm) [inline execute/transport.go:270]: Thank you for all these comments!

**Review by wolffcm — COMMENTED** [review]
This seems reasonable to me, but I'm not clear on the practical benefit of these changes, can you answer that for me?

**Comment by jsternberg** [comment]
The main purpose was to help with extensibility and composability. At the present moment, if we want to add a new message type to the system, then either a change to the `Transformation` interface or using a `typecast` to something like:

```
type ProcessViewTransformation interface {
    Transformation
    ProcessView(...) error
}
```

One of these two is required because the consecutive transport masquerades as a transformation and needs to forward the message through function calls. This makes the `Transformation` interface very heavy and prone to break. The second version also doesn't work very well when wrapping another type because typecasts can't be forwarded to wrapped types. This simplifies the requirements for a transformation by just saying "If you implement this one method you're good" and then each transformation that implements that method can decide to just ignore messages they don't understand or perform some default actions. This makes delegation and wrapping easier.

In the next commit after this one, I add some new message types that execkit transformations will both produce and handle and the transformation adapter is meant to communicate between these two message handling paradigms. I'll push up that commit so you can see what I mean.

**Comment by jsternberg** [comment]
Here's the follow up commit to this one: https://github.com/influxdata/flux/commit/c23c74a2f7b35bd64956d6106c412d7947bfd761.

**Comment by wolffcm** [comment]
Thinking this over a bit more and looking at the follow up commit that you linked, I have some concerns. It seems like it adds complexity to the query execution layer--an area I already find more than a bit confusing. I suspect that there's some good ideas here, but I think this should be opened up to a broader discussion with the team, and I don't know that we have the bandwidth right now.

@nathanielc What are your thoughts on this, with regards to prioritization or otherwise?

**Pull request closed without merging**